control-plane-operator/hostedclusterconfigoperator/controllers/resources/resources_test.go

```diff
@@ -365,3 +366,198 @@ func TestReconcileUserCertCABundle(t *testing.T) {
+			},
+		}
+	}
+
+	verifyIngressControllersRemoved := func(g *WithT, c client.Client) {
+		ingressControllers := &operatorv1.IngressControllerList{}
+		err := c.List(context.Background(), ingressControllers)
+		g.Expect(err).ToNot(HaveOccurred())
+		g.Expect(len(ingressControllers.Items)).To(Equal(0))
+	}
+
+	serviceLoadBalancer := func(name string) client.Object {
+		return &corev1.Service{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      name,
+				Namespace: "default",
+			},
+			Spec: corev1.ServiceSpec{
+				Type: corev1.ServiceTypeLoadBalancer,
+			},
+		}
+	}
+
+	clusterIPService := func(name string) client.Object {
+		return &corev1.Service{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      name,
+				Namespace: "default",
+			},
+			Spec: corev1.ServiceSpec{
+				Type: corev1.ServiceTypeClusterIP,
+			},
+		}
+	}
+
+	verifyServiceLoadBalancersRemoved := func(g *WithT, c client.Client) {
+		services := &corev1.ServiceList{}
+		err := c.List(context.Background(), services)
+		g.Expect(err).ToNot(HaveOccurred())
+		for _, svc := range services.Items {
+			g.Expect(svc.Spec.Type).ToNot(Equal(corev1.ServiceTypeLoadBalancer))
+		}
+	}
+
+	verifyServiceExists := func(name string, g *WithT, c client.Client) {
+		service := clusterIPService(name)
+		err := c.Get(context.Background(), client.ObjectKeyFromObject(service), service)
+		g.Expect(err).ToNot(HaveOccurred())
+	}
+
+	pv := func(name string) client.Object {
+		return &corev1.PersistentVolume{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: name,
+			},
+		}
+	}
+
+	pvc := func(name string) client.Object {
+		return &corev1.PersistentVolumeClaim{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      name,
+				Namespace: "default",
+			},
+		}
+	}
+
+	verifyPVCsRemoved := func(g *WithT, c client.Client) {
+		pvcs := &corev1.PersistentVolumeClaimList{}
+		err := c.List(context.Background(), pvcs)
+		g.Expect(err).ToNot(HaveOccurred())
+		g.Expect(len(pvcs.Items)).To(Equal(0))
+	}
+
+	tests := []struct {
+		name     string
+		existing []client.Object
+		verify   func(*WithT, client.Client)
+	}{
+		{
+			name: "no existing resources",
+		},
+		{
+			name: "image registry with storage",
+			existing: []client.Object{
+				managedImageRegistry(),
+			},
+			verify: verifyImageRegistryConfig,
+		},
+		{
+			name: "existing ingress controller",
+			existing: []client.Object{
+				ingressController("default"),
+				ingressController("foobar"),
+			},
+			verify: verifyIngressControllersRemoved,
+		},
+		{
+			name: "existing service load balancers",
+			existing: []client.Object{
+				serviceLoadBalancer("foo"),
+				serviceLoadBalancer("bar"),
+				clusterIPService("baz"),
+			},
+			verify: func(g *WithT, c client.Client) {
+				verifyServiceLoadBalancersRemoved(g, c)
+				verifyServiceExists("baz", g, c)
+			},
+		},
+		{
+			name: "existing pv/pvc",
+			existing: []client.Object{
+				pv("foo"), pvc("foo"),
+				pv("bar"), pvc("bar"),
+			},
+			verify: verifyPVCsRemoved,
+		},
```

Comment:
Could we add a pod test here and only put the pods into the `uncachedClient`? That aspect is really important